modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/modules.tsx

```diff
@@ -0,0 +1,135 @@
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import {loadFDSCellRenderer} from '@liferay/client-extension-web';
+import React, {RefObject} from 'react';
+
+import type {FDSCellRendererArgs} from '@liferay/js-api/data-set';
+
+export async function getLiferayJsModule(moduleURL: string): Promise<any> {
+	if (moduleURL.includes(' from ')) {
+		const parts = moduleURL.split(' from ');
+		const moduleName = parts[1].trim();
+
+		let symbolName = parts[0].trim();
+
+		if (
+			symbolName !== 'default' &&
+			(!symbolName.startsWith('{') || !symbolName.endsWith('}'))
+		) {
+			throw new Error(`Invalid module URL: ${moduleURL}`);
+		}
+
+		if (symbolName.startsWith('{')) {
+			symbolName = symbolName.substring(1, symbolName.length - 1).trim();
+		}
+
+		try {
+			const renderer = await loadFDSCellRenderer(moduleName, symbolName);
+
+			return Promise.resolve((props: any) => {
+				return React.createElement(
+
+					// @ts-ignore
+
+					class extends React.Component {
+						private readonly ref: RefObject<HTMLDivElement>;
+						private readonly args: FDSCellRendererArgs;
+
+						constructor(props: {value: any}) {
+							super(props);
+
+							this.ref = React.createRef();
+							this.args = {
+								value: props.value,
+							};
+						}
+
+						render() {
+							setTimeout(() => {
+								if (this.ref.current) {
+									this.ref.current.appendChild(
+										renderer(this.args)
+									);
+								}
+							}, 0);
+
+							return <div ref={this.ref}></div>;
+						}
+					},
+					props
+				);
+			});
+		}
+		catch (error) {
+			console.error(error);
+
+			// Rejecting the promise enters an infinite loop
+
+			return Promise.resolve(() => (
+				<div className="custom-component">component load failed</div>
+			));
+		}
+	}
+	else {
+		return new Promise((resolve, reject) => {
+			Liferay.Loader.require(
+				moduleURL,
+				(jsModule) => resolve(jsModule.default || jsModule),
+				(error) => reject(error)
+			);
+		});
+	}
+}
+
+export function getFakeJsModule() {
+	return new Promise((resolve) => {
+		setTimeout(
+			() =>
+				resolve(() => (
+					<div className="custom-component">
+						fakely fetched component
+					</div>
+				)),
+			3000
+		);
+	});
+}
+
+export const getJsModule = (Liferay as any).Loader?.require
+	? getLiferayJsModule
+	: getFakeJsModule;
+
+export const fetchedJsModules: Array<{component: any; url: string}> = [];
+
+export async function getComponentByModuleURL(url: string): Promise<any> {
+	return new Promise((resolve, reject) => {
+		const foundModule = fetchedJsModules.find((cr) => cr.url === url);
+
+		if (foundModule) {
+			resolve(foundModule.component);
+		}
```

Comment:
I guess this cache is useful for AMD based loading but not for the ES Browser Modules as browser takes care. Another advantage of keeping this file only for AMD-based loading is that we can separate the logic and let our ES-based dynamic imports manage this

Reply:
If it is useless for ESM as a caching mechanism it is useless for AMD as well, because both models are exactly equivalent. In the sense that [this commentary](https://github.com/liferay-frontend/liferay-portal/pull/3119#discussion_r1138245729) applies to AMD too.

Reply:
One thing I noticed in the FDS code is that sometimes it made the requests for each cell in parallel, which lead to new imports being raised when the first one hadn't finished yet. The cache could probably be added for that in first place, but I'm not sure, since I don't know the exact details on how FDS rendering is done (or wants to be done).

All I know is that it is making way too many assumptions as to how async executions happen and are queued, because if I changed that workflow I got weird behaviors. For example, if [this setTimeout](https://github.com/liferay-frontend/liferay-portal/pull/3119/files#diff-7264773118929c273c70b7252decc907b71b2ebb83e5fdf71f9c76fc2c3cacc0R60-R66) is incremented to 1, it will be deferred when queued and will lead to an infinite loop if the cell renderer doesn't load and is rejected (that's why I [changed the reject](https://github.com/liferay-frontend/liferay-portal/pull/3119/files#diff-7264773118929c273c70b7252decc907b71b2ebb83e5fdf71f9c76fc2c3cacc0R78) too to always return a valid component and be safe).

Once this is settled and agreed we should definitely investigate the rendering lifecycle and make it more robust.